NFC support for SetUpCodePairer and AutoCommissioner

anush-apple opened the pull request:
This change is dependent on https://github.com/project-chip/connectedhomeip/pull/40780. 

#### Summary

- Support for NFC discovery in SetUpCodePairer
- AutoCommissioner support for two phase stage set for NFC based commissioning
- Added a new kUnpoweredStageComplete to mark end phase1 of commissioning. This allows Commissioner to remember this for the device and use the new API (ContinueCommissioningAfterUnpoweredPhaseComplete) to continue commissioning post powerup of the device.

#### Related issues

#40779


#### Testing

Feature is not enabled by default. Bringing in code in smaller pieces for easier review

#### Readability checklist

The checklist below will help the reviewer finish PR review in time and keep the
code readable:

-   [x] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html#title-formatting)
-   [x] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [x] PR size is short
-   [x] Try to avoid "squashing" and "force-update" in commit history
-   [x] CI time didn't increase

See: [Pull Request Guidelines](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html)

anush-apple committed:
NFC support for SetUpCodePairer and AutoCommissioner

- Support for NFC discovery in SetUpCodePairer
- AutoCommissioner support for two phase stage set for NFC based
  commissioning
- Added a new kUnpoweredStageComplete to mark end phase1 of
  commissioning. This allows Commissioner to remember this for the
  device and use the new API (ContinueCommissioningAfterUnpoweredPhaseComplete)
  to continue commissioning post powerup of the device.

OlivierGre commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -487,7 +487,22 @@ CommissioningStage AutoCommissioner::GetNextCommissioningStageInternal(Commissio
+        {
+            // If there is no secure session, means commissioning has been continued after the unpowered phase was completed
+            // In such case, move on setup the CASE session over operational network
+            if (!mCommissioneeDeviceProxy->GetSecureSession().HasValue())
```
Can you please explain me how it works and what is mCommissioneeDeviceProxy?
When doing NFC commissioning with or without power, there is a PASE secure session setup with the tag. This is probably not what you are talking about here.

OlivierGre commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -1071,6 +1085,46 @@ DeviceCommissioner::ContinueCommissioningAfterDeviceAttestation(DeviceProxy * de
+#if CHIP_DEVICE_CONFIG_ENABLE_NFC_BASED_COMMISSIONING
+CHIP_ERROR DeviceCommissioner::ContinueCommissioningAfterUnpoweredPhaseComplete(NodeId remoteDeviceId)
+{
+    MATTER_TRACE_SCOPE("continueCommissioningDeviceAfterUnpoweredPhase", "DeviceCommissioner");
```
ContinueCommissioningAfterUnpoweredPhaseComplete

OlivierGre commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -578,17 +597,33 @@ CHIP_ERROR AutoCommissioner::StartCommissioning(DeviceCommissioner * commissione
         return CHIP_ERROR_INVALID_ARGUMENT;
     }
 
+#if CHIP_DEVICE_CONFIG_ENABLE_NFC_BASED_COMMISSIONING
```
Please explain how the current code in StartCommissioning() is used for NFC commissioning without power and how it is used with ContinueCommissioningAfterUnpoweredPhaseComplete. It is not clear to me what is the role of each one.

OlivierGre commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -508,6 +523,10 @@ CommissioningStage AutoCommissioner::GetNextCommissioningStageInternal(Commissio
         return CommissioningStage::kCleanup;
+#if CHIP_DEVICE_CONFIG_ENABLE_NFC_BASED_COMMISSIONING
+    case CommissioningStage::kUnpoweredPhaseComplete:
+        return CommissioningStage::kCleanup; // End commissioning (phase 1)
```
I'm wondering if the Commissioner will not lose some useful things about the device that has been pre commissioned over NFC.

I was thinking that we would explicitely have function to save and restore the commissioning context.

I was thinking to 2 different ways of doing NFC commissioning without power:

- The user does a pre commissioning of a device.
- He goes and install the device
- He confirms in the app that he has installed the device
- The 2nd commissioning phase is executed.
In this case no need to save and restore the commissioning data because the user do a full commissioning of a single device.

The other solution is:
- Pre commissioning of a device 1
- save commissioning data and clean up
- Pre commissioning of a device 2
- save commissioning data and clean up
- Pre commissioning of a device 3
- save commissioning data and clean up

Later on:
- Restore commissioning data of device 1
- Perform the 2nd commissioning phase on the operational network. Commissioning is completed.
- Restore commissioning data of device 2
- Perform the 2nd commissioning phase on the operational network. Commissioning is completed.
- Restore commissioning data of device 3
- Perform the 2nd commissioning phase on the operational network. Commissioning is completed.

OlivierGre commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -487,7 +487,22 @@ CommissioningStage AutoCommissioner::GetNextCommissioningStageInternal(Commissio
+
+            // If the transport is NFC, move to unpowered phase complete to end the first phase of commissioning
+            auto transportType = mCommissioneeDeviceProxy->GetSecureSession().Value()->AsSecureSession()->GetPeerAddress().GetTransportType();
+            return (transportType == Transport::Type::kNfc || mDeviceCommissioningInfo.general.isCommissioningWithoutPower) ? CommissioningStage::kUnpoweredPhaseComplete : CommissioningStage::kFindOperationalForStayActive;
```
This condition is not correct for NFC commissioning with power. Because of this line, you will reach the stage kUnpoweredPhaseComplete which is not expected.

anush-apple committed:
Code review comments

restyled-commits committed:
Restyled by clang-format

OlivierGre commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -1071,6 +1085,46 @@ DeviceCommissioner::ContinueCommissioningAfterDeviceAttestation(DeviceProxy * de
+                    ChipLogValueX64(remoteDeviceId));
+
+    MATTER_LOG_METRIC_BEGIN(kMetricDeviceCommissioningOperationalSetup);
+    CHIP_ERROR err = mDefaultCommissioner->StartCommissioning(this, device);
```
Ok. I think that we didn't have exactly the same focus and this explains some differences in the code.
There are 2 different use cases:
- there is the case of the installation of a single device via NFC commissioning.
- Then, there is the case of pre commissioning several devices via NFC and then executing the second commissioning phase later on.

In my PRs, I was focussing on the first use case. I was considering that having it in Matter 1.5 would be a good achievement and I was seeing the 2nd use case for Matter 1.6.

I see that you dived in with a focus that is more on the 2nd use case (which is fine, no problem with that). This explains why I didn't have code to start commissioning for the 2nd phase.

anush-apple replied:
This change allows for commissioner to support your first use-case. They just call ContinueCommissioningAfterUnpoweredPhaseComplete immediately and it will maintain the same behavior

anush-apple commented:
Android build failure is expected since this is dependent on https://github.com/project-chip/connectedhomeip/pull/40780

anush-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -578,17 +600,36 @@ CHIP_ERROR AutoCommissioner::StartCommissioning(DeviceCommissioner * commissione
+    // commissioning post unpowered phase, allow proceeding if the stage is evict secure CASE sessions
+    if (proxy == nullptr ||
+        (!proxy->GetSecureSession().HasValue() &&
+         commissioner->GetCommissioningStage() != CommissioningStage::kEvictPreviousCaseSessions))
```
I did that first and I really did not like the readability due to maintaining the closing ')'

anush-apple replied:
Updated

anush-apple committed:
Code review fixes

restyled-commits committed:
Restyled by clang-format

anush-apple committed:
Renamed API to support existing pattern

- AutoCommissioner has SkipCommissioningComplete flags. Renamed
  API to stage name to allow using it if flag is a set as well

anush-apple committed:
Fixed tracing string to match API name

anush-apple committed:
Merge branch 'master' into darwin-nfc-support-part2

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -487,7 +487,25 @@ CommissioningStage AutoCommissioner::GetNextCommissioningStageInternal(Commissio
+#if CHIP_DEVICE_CONFIG_ENABLE_NFC_BASED_COMMISSIONING
+    {
+        // If there is no secure session, means commissioning has been continued after the unpowered phase was completed
+        // In such case, move on setup the CASE session over operational network
```
```
        // If there is no secure session, that means commissioning has been continued after the unpowered phase was completed.
        // In that case, move on to set up the CASE session over operational network.
```

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -487,7 +487,25 @@ CommissioningStage AutoCommissioner::GetNextCommissioningStageInternal(Commissio
+        // If the transport is NFC, move to unpowered phase complete to end the first phase of commissioning
+        auto transportType =
+            mCommissioneeDeviceProxy->GetSecureSession().Value()->AsSecureSession()->GetPeerAddress().GetTransportType();
+        return (transportType == Transport::Type::kNfc && mDeviceCommissioningInfo.general.isCommissioningWithoutPower)
```
Why is this logic NFC-specific?  The whole point of isCommissioningWithoutPower is that it's transport-independent, no?  If we end up with another transport that can handle no-power commissioning, it should ideally just work...

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -508,6 +526,10 @@ CommissioningStage AutoCommissioner::GetNextCommissioningStageInternal(Commissio
     case CommissioningStage::kSendComplete:
         return CommissioningStage::kCleanup;
+#if CHIP_DEVICE_CONFIG_ENABLE_NFC_BASED_COMMISSIONING
+    case CommissioningStage::kUnpoweredPhaseComplete:
```
Again, does this need to be ifdefed on NFC?

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -578,7 +600,14 @@ CHIP_ERROR AutoCommissioner::StartCommissioning(DeviceCommissioner * commissione
+    // commissioning post unpowered phase, allow proceeding if the stage is evict secure CASE sessions
+    if (proxy == nullptr ||
+        (!proxy->GetSecureSession().HasValue()
+#if CHIP_DEVICE_CONFIG_ENABLE_NFC_BASED_COMMISSIONING
```
Again, does this need to be NFC-specific?

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -587,8 +616,15 @@ CHIP_ERROR AutoCommissioner::StartCommissioning(DeviceCommissioner * commissione
-    auto transportType =
-        mCommissioneeDeviceProxy->GetSecureSession().Value()->AsSecureSession()->GetPeerAddress().GetTransportType();
+    // When commissioning is started after unpowered phase, there will be no secure session since the CASE session first needs to be
+    // setup over operational network. Hence assume transport is UDP in such case.
```
"set up" and "in this case"

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -632,6 +636,16 @@ void DeviceCommissioner::ReleaseCommissioneeDevice(CommissioneeDeviceProxy * dev
+    if (readerTransport)
+    {
+        ChipLogProgress(Controller, "Stopping discovery of all NFC tags");
+        readerTransport->StopDiscoveringTags();
```
I know this tries to do the same thing as BLE, but it's really not ideal:

1. If the commissionee being released is not actually on the NFC transport (we can have multiple commissionee devices around, sort of!), this will do the wrong thing.  Note that BLE does check for BLE transport.
2. Ideally we would clearly communicate that whatever handle we have to NFC is now released and NFC would shut itself down once all the handles are gone, or something, instead of having this global "nuke everything" setup.

Is there a followup tracking something more reasonable here?

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -1071,6 +1085,49 @@ DeviceCommissioner::ContinueCommissioningAfterDeviceAttestation(DeviceProxy * de
     return CHIP_NO_ERROR;
 }
 
+#if CHIP_DEVICE_CONFIG_ENABLE_NFC_BASED_COMMISSIONING
```
Does this need to be NFC-specific?  In theory this all works for any transport that supports an unpowered phase, right?

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -1071,6 +1085,49 @@ DeviceCommissioner::ContinueCommissioningAfterDeviceAttestation(DeviceProxy * de
+    // on the operational network
+    mCommissioningStage = CommissioningStage::kEvictPreviousCaseSessions;
+
+    // Setup device being commissioned
```
"set up"

bzbarsky-apple commented on src/controller/CommissioningDelegate.h:
```diff
@@ -89,6 +89,9 @@ enum CommissioningStage : uint8_t
     kRemoveThreadNetworkConfig,       ///< Remove Thread network config.
     kConfigureTCAcknowledgments,      ///< Send SetTCAcknowledgements (0x30:6) command to the device
     kCleanup,                         ///< Call delegates with status, free memory, clear timers and state/
+#if CHIP_DEVICE_CONFIG_ENABLE_NFC_BASED_COMMISSIONING
```
Again, does this need to be ifdefed?

bzbarsky-apple commented on src/controller/SetUpCodePairer.cpp:
```diff
@@ -300,6 +315,66 @@ CHIP_ERROR SetUpCodePairer::StopDiscoveryOverWiFiPAF()
+    }
+    else
+    {
+        mWaitingForDiscovery[kNFCTransport] = true;
```
Is StartDiscoveringTagMatchingAddress allowed to synchronously notify the delegate when it returns CHIP_NO_ERROR?

The API docs don't say, and if it does that this logic will not work correctly.

bzbarsky-apple commented on src/controller/SetUpCodePairer.cpp:
```diff
@@ -300,6 +315,66 @@ CHIP_ERROR SetUpCodePairer::StopDiscoveryOverWiFiPAF()
+    const SetupDiscriminator connDiscriminator(payload.discriminator);
+    VerifyOrReturnValue(!connDiscriminator.IsShortDiscriminator(), CHIP_ERROR_INVALID_ARGUMENT,
+                        ChipLogError(Controller, "Error, Long discriminator is required"));
+    chip::Nfc::NFCTag::Identifier identifier  = { .discriminator = payload.discriminator.GetLongValue() };
```
Why the various `chip::` prefixes here?  This stuff is all inside `namespace chip`....
